bindings/nodejs/hawktracer_client_nodejs.cpp

```diff
@@ -0,0 +1,185 @@
+#include "hawktracer_client_nodejs.hpp"
+
+#include <iostream>
+#include <utility>
+
+namespace HawkTracer
+{
+namespace Nodejs
+{
+
+Object Client::init_bindings(class Env env, Object exports)
+{
+    HandleScope scope(env);
+
+    Function constructor = DefineClass(
+        env,
+        "HawkTracerClient",
+        {
+            InstanceMethod("start", &Client::start),
+            InstanceMethod("onEvents", &Client::set_on_events),
+            InstanceMethod("stop", &Client::stop)
+        });
+    Persistent(constructor).SuppressDestruct();
+    exports.Set("HawkTracerClient", constructor);
+    return exports;
+}
+
+Client::Client(const CallbackInfo &info)
+    : ObjectWrap<Client>(info)
+{
+    _source = info[0].As<String>();
+}
+
+Client::~Client()
+{
+    _stop();
+}
+
+Value Client::start(const CallbackInfo &info)
+{
+    _context_holder->context = ClientContext::create(
+        _source,
+        [this](std::unique_ptr<std::vector<parser::Event>> data, ClientContext::ConsumeMode consume_mode)
+        {
+            return handle_events(std::move(data), consume_mode);
+        });
+    return Boolean::New(info.Env(), static_cast<bool>(_context_holder->context));
+}
+
+void Client::stop(const CallbackInfo &)
+{
+    _stop();
+}
+
+void Client::_stop()
+{
+    std::lock_guard<std::mutex> lock{_callback_mutex};
+    if (_callback) {
+        _callback->function.Abort();
+        _callback->function.Release();
+        _callback.reset();
+    }
+    else {
+        delete _context_holder;
+    }
+}
+
+void Client::set_on_events(const CallbackInfo &info)
+{
+    std::lock_guard<std::mutex> lock{_callback_mutex};
+    if (_callback) {
+        // existing _context_holder will be deleted by the finalizer of existing _callback.function
+        _context_holder = new ContextHolder{std::move(*_context_holder)};
+        _callback->function.Release();
+    }
+    _callback.reset(new ThreadSafeFunctionHolder{
+        ThreadSafeFunction::New(info.Env(),
+                                info[0].As<Napi::Function>(),
+                                "HawkTracerClientOnEvent",
+                                1,
+                                1,
+                                [](class Env, decltype(_context_holder) context_holder)
+                                {
+                                    delete context_holder;
+                                },
+                                _context_holder)});
+}
+
+// This method is called from reader thread, while all other methods are called from js main thread.
+std::unique_ptr<std::vector<parser::Event>>
+Client::handle_events(std::unique_ptr<std::vector<parser::Event>> events, ClientContext::ConsumeMode consume_mode)
+{
+    std::lock_guard<std::mutex> lock{_callback_mutex};
+    if (!_callback) {
+        return events;
```

Comment:
Ring buffer seems a good idea for the time before we have event handler. But I'm not sure if losing events in between event handler calls is a good idea. If there are too many messages, the application should be decide how to deal with it.
As application developer, I would prefer having some kind of steam mechanism (such as rxjs) and I can easily create a stream from the events. So I would just leave it as it is.
What do you think? Do you think it's worth having it as a ring buffer just for the time before `onEvents()` call?

Reply:
I think in usual scenario, start() and onEvents() will be called closely in time. I'll skip implementing ring buffer, since it's not a critical feature.